// disaster-response-backend/routes/users.js
const express = require('express');
const router = express.Router();

// Basic user routes for future implementation
// For now, we'll keep it simple without authentication

// Get user profile (placeholder)
router.get('/profile', (req, res) => {
  res.json({
    message: 'User profile endpoint - authentication required',
    user: {
      id: 'demo-user',
      name: 'Demo User',
      email: 'demo@example.com',
      role: 'citizen',
      location: 'Ghaziabad, UP',
      notifications: {
        email: true,
        sms: true,
        push: true
      }
    }
  });
});

// Update user preferences (placeholder)
router.patch('/preferences', (req, res) => {
  const { notifications, alertRadius, emergencyContacts } = req.body;
  
  res.json({
    message: 'User preferences updated successfully',
    preferences: {
      notifications,
      alertRadius,
      emergencyContacts
    }
  });
});

// Get user's alert history (placeholder)
router.get('/alerts', (req, res) => {
  res.json({
    message: 'User alert history',
    alerts: [
      {
        id: 'alert-1',
        type: 'earthquake',
        severity: 'warning',
        acknowledged: true,
        timestamp: new Date(Date.now() - 24 * 60 * 60 * 1000)
      }
    ]
  });
});

module.exports = router;
